src/lib.rs

```diff
@@ -20,6 +20,79 @@ mod pippenger;
 mod scalar_mul;
 mod schnorr;
 
+use std::array::TryFromSliceError;
+
+use thiserror::Error;
+
+// There are no WasmErrors on native
+#[cfg(feature = "native")]
+#[derive(Debug, Error)]
+pub enum WasmError {}
+
+#[cfg(not(feature = "native"))]
+#[derive(Debug, Error)]
+pub enum WasmError {
+    #[error("Trying to call {name} resulted in an error")]
+    FunctionCallFailed {
+        name: String,
+        source: wasmer::RuntimeError,
+    },
+    #[error("Could not find function export named {name}")]
+    InvalidExport {
+        name: String,
+        source: wasmer::ExportError,
+    },
+    #[error("No value available when value was expected")]
+    NoValue,
+    #[error("Value expected to be i32")]
+    InvalidI32,
+    #[error("Could not convert value {value} from i32 to u32")]
+    InvalidU32 {
+        value: i32,
+        source: std::num::TryFromIntError,
+    },
+    #[error("Value expected to be 0 or 1 representing a boolean")]
+    InvalidBool,
+    #[error("Failed to get pointer")]
+    InvalidPointer { source: TryFromSliceError },
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("The value {0} overflows in the pow2ceil function")]
+    Pow2CeilOverflow(u32),
+    #[error("Could not convert schnorr")]
+    SchnorrConvert(Vec<u8>),
+    #[error("Could not slice schnorr")]
+    SchnorrSlice { source: TryFromSliceError },
+    #[error("Could not slice field element")]
+    FieldElementSlice { source: TryFromSliceError },
+    #[error("Expected a Vec of length {0} but it was {1}")]
+    FieldToArray(usize, usize),
+
+    #[error("Missing rest of output. Tried to get byte {0} but failed")]
+    MissingOutput(usize),
+    #[error("Missing leaf to check membership for")]
+    MissingLeaf,
+    #[error("Missing index for leaf")]
+    MissingLeafIndex,
+    #[error("Missing `x` component for public key")]
+    MissingPublicKeyX,
+    #[error("Missing `y` component for public key")]
+    MissingPublicKeyY,
+    #[error("Missing rest of signature. Tried to get byte {0} but failed")]
+    MissingSignature(usize),
+    #[error("Missing rest of {0} component for public key. Tried to get byte {1} but failed")]
+    MissingPublicKey(&'static str, usize),
+    #[error("Keccak256 has not yet been implemented")]
+    Keccak256,
+    #[error("AES has not yet been implemented")]
+    Aes,
```

Comment:
This doesn't seem possible because that's an enum and if I try something like:
```
#[error(transparent)]
UnsupportedBlackBoxFunc(#[from] OpcodeResolutionError::UnsupportedBlackBoxFunc()),
```
It says
> try using the variant's enum: `acvm::OpcodeResolutionError`, `crate::Error`, `crate::OpcodeResolutionError`

which is the incorrect way to handle this error since it can't be anything from the enum.